sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -407,260 +408,284 @@ class GpuSequenceMeta(
-    //     since when size < 0, cudf will not generate sequence
-    // Second, calculate size = if(sizeWithNegative < 0) 0 else sizeWithNegative
-    // Third, if (start == stop && step == 0), let size = 1.
-    withResource(GpuSequenceUtil.numberScalar(dt, 1)) { one =>
-      withResource(GpuSequenceUtil.numberScalar(dt, 0)) { zero =>
+  override def withTimeZone(timeZoneId: String): TimeZoneAwareExpression =
+    copy(timeZoneId = Some(timeZoneId))
 
-        val (sizeWithNegative, diffHasZero) = withResource(stop.sub(start)) { difference =>
-          // sizeWithNegative=floor((stop-start)/step)+1
-          val sizeWithNegative = withResource(difference.floorDiv(step)) { quotient =>
-            quotient.add(one)
-          }
-          val tmpDiffHasZero = closeOnExcept(sizeWithNegative) { _ =>
-            difference.equalTo(zero)
-          }
-          (sizeWithNegative, tmpDiffHasZero)
-        }
+  override def children: Seq[Expression] = Seq(start, stop) ++ stepOpt
 
-        val tmpSize = closeOnExcept(diffHasZero) { _ =>
-          // tmpSize = if(sizeWithNegative < 0) 0 else sizeWithNegative
-          withResource(sizeWithNegative) { _ =>
-            withResource(sizeWithNegative.greaterOrEqualTo(zero)) { pred =>
-              pred.ifElse(sizeWithNegative, zero)
-            }
-          }
-        }
+  override def nullable: Boolean = children.exists(_.nullable)
 
-        // when start==stop && step==0, size will be 0.
-        // but we should change size to 1
-        withResource(tmpSize) { tmpSize =>
-          withResource(diffHasZero) { diffHasZero =>
-            step match {
-              case stepScalar: Scalar =>
-                withResource(ColumnVector.fromScalar(stepScalar, rows)) { stepV =>
-                  withResource(stepV.equalTo(zero)) { stepHasZero =>
-                    withResource(diffHasZero.and(stepHasZero)) { predWithZero =>
-                      predWithZero.ifElse(one, tmpSize)
-                    }
-                  }
-                }
-              case _ =>
-                withResource(step.equalTo(zero)) { stepHasZero =>
-                  withResource(diffHasZero.and(stepHasZero)) { predWithZero =>
-                    predWithZero.ifElse(one, tmpSize)
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    // 1 Compute the array size for all the rows
+    val (startCol, sizeCol, stepCol, nonNullPredOpt) =
+      withResource(columnarEvalToColumn(start, batch)) { startGpuCol =>
+        withResource(columnarEvalToColumn(stop, batch)) { stopGpuCol =>
+          withResource(stepOpt.map(columnarEvalToColumn(_, batch))) { stepColOpt =>
+            // Filter out the nulls before any computation, since cudf 'sequences' does
+            // not allow nulls.(Pls refer to https://github.com/rapidsai/cudf/issues/10012)
+            // Then no need to do computation for the null rows.
```

Comment:
I didn't get the impression that we were actually supposed to `filter` the nulls out here.  Per the comment at https://github.com/rapidsai/cudf/issues/10012#issuecomment-1010302972 I was under the impression we're simply doing a logical AND of the start/stop/step validity masks to determine which rows will be valid in the output.  For every row that is not going to be valid, we update the `sizeCol` to be zero for those corresponding entries (can build up a column view with the AND'd validity mask and use `replaceNulls` to do this easily).  Then we execute the sequence with the validity masks stripped off, and then use the AND'd validity mask as the mask of the final resulting column.

That should preclude the need to do any fancy prefix sum scan, gather shenanigans, etc.

Reply:
Thanks for review.
Personally I like the "filter-out nulls" way which follows the processing in Spark, and the logic seems less cudf specific. 
But I am trying your suggestion locally, still struggling with many failures in tests due to different output between CPU and GPU, will continue it next week.

Reply:
Eventually, I have made it works. Updated